### PR TITLE
jsonschema containers.json: allow empty sections

### PR DESCRIPTION
**Pull request by MartinBasti**
- [x] PoC with "platforms" attribute
- [x] Allow other attributes (sections) to be empty
- [x] Check if other attributes are handled well when set to `None`
- [x] Tests

Hello,
I've created a PoC implementation. Please check if I placed tests to the proper location, if tests are written in expected form, if I missed anything, please check commit messages, code style, etc.... I want to fix my newbie errors as soon as possible :-)

Also it looks that other attributes can be used as empty sections, I'd go ahead and fix it everywhere in `containers.json`, is that OK?

Martin

### CODE REVIEW

File: tests/test_source.py [hunk 1]
```diff
@@ -79,3 +87,58 @@ def test_broken_source_config_file(self):
...
+        platforms:
+            only: s390x
+        """,
+
+        """\
+        platforms:
+            not: s390x
+        """,
+
+        # TODO: schema probably should not allow to use both keywords
```
Comment (twaugh) [inline tests/test_source.py]: Actually the schema deliberately allows this, but it should document that "only" is processed before "not". The reason is that this allows a temporary exclusion of an architecture, e.g.:
```
only: abc def ghi jkl mno
# not yet working on ghi, bug #.....
not: ghi
```

**Comment by twaugh** [comment]
> Also it looks that other attributes can be used as empty sections, I'd go ahead and fix it everywhere in containers.json, is that OK?

Yes, we ought to be able to handle sections whose keys are all commented-out.

File: tests/test_source.py (hunk 1 above)
Comment (MartinBasti) [reply tests/test_source.py]: Thanks for clarification, TODO removed.

Test for `get_platforms_in_limits` added

**Comment by MartinBasti** [comment]
Do we want also support for:

```
platforms:
   only:
   # - x86_64
   not: ppc64le
```

it needs some changes in `get_platforms_in_limits` ?

**Comment by twaugh** [comment]
Agreed, I think we should not allow that.

**Comment by MartinBasti** [comment]
I did what was requested in ticket, but so far it looks that it is not enough. If I use config values that violates schema, instead of schema error I got weird error (see traceback). It looks that config is not validated early enough by schema. (BTW: I have unit test for that particular case and schema works as expected)

```
cat container.yaml
---
platforms: not_an_object
```

orchestrator.log
```
2018-10-15 16:57:22,238 platform:- - atomic_reactor.plugin - DEBUG - Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugin.py", line 239, in run
    plugin_response = plugin_instance.run()
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugins/pre_check_and_set_platforms.py", line 85, in run
    return get_platforms_in_limits(self.workflow, enabled_platforms)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 1525, in get_platforms_in_limits
    excluded_platforms = set(make_list(data['platforms'].get('not', [])))
AttributeError: 'str' object has no attribute 'get'
2018-10-15 16:57:22,238 platform:- - atomic_reactor.plugin - ERROR - plugin 'check_and_set_platforms' raised an exception: AttributeError("'str' object has no attribute 'get'",)

-----

2018-10-15 16:57:22,344 platform:- - atomic_reactor.plugin - DEBUG - plugin 'remove_worker_metadata' finished in 0s
Traceback (most recent call last):
  File "/usr/bin/atomic-reactor", line 11, in <module>
    load_entry_point('atomic-reactor==1.6.35', 'console_scripts', 'atomic-reactor')()
  File "/usr/lib/python2.7/site-packages/atomic_reactor/cli/main.py", line 309, in run
    cli.run()
  File "/usr/lib/python2.7/site-packages/atomic_reactor/cli/main.py", line 292, in run
    args.func(args)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/cli/main.py", line 95, in cli_inside_build
    substitutions=args.substitute)
  File "/usr/lib/python2.7/site-packages/atomic_reactor/inner.py", line 546, in build_inside
    build_result = dbw.build_docker_image()
  File "/usr/lib/python2.7/site-packages/atomic_reactor/inner.py", line 429, in build_docker_image
    prebuild_runner.run()
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugin.py", line 279, in run
    raise PluginFailedException(msg)
atomic_reactor.plugin.PluginFailedException: plugin 'check_and_set_platforms' raised an exception: AttributeError("'str' object has no attribute 'get'",)
```

The reason is that function `get_platforms_in_limits` loads `container.yaml` config without any validation. IMO this is not a correct behavior, there should be a single object responsible for parsing and loading container config (I assumed that this is `SourceConfig`, the only object that uses the particular schema validation).
Currently I don't know exactly how to glue it together, i.e. having only `SourceConfig` for container config data. I can use it as instance directly in `get_platforms_in_limits`, but it doesn't look like the best solution to me. Config files should be validated as early as possible.

**Comment by MartinBasti** [comment]
After discussion with @rcerven this may be a solution to enforce validation
```diff
-    build_file_dir = workflow.source.get_build_file_path()[1]
-    container_path = os.path.join(build_file_dir, REPO_CONTAINER_CONFIG)
-    if os.path.exists(container_path):
-        with open(container_path) as f:
-            data = yaml.safe_load(f)
+  data = workflow.source.config
```

**Comment by MartinBasti** [comment]
Ready for review :tada:

File: atomic_reactor/source.py [hunk 2]
```diff
@@ -37,16 +37,21 @@ class SourceConfig(object):
...
             except Exception:
-                logger.exception("Failed to load and validate source config YAML from " + file_path)
+                logger.exception(
+                    "Failed to load and validate source config YAML from %s",
+                    self.file_path
+                )
                 raise
 
-        self.autorebuild = self.data.get('autorebuild', {})
+        self.autorebuild = self.data.get('autorebuild') or {}
```
[the remark this answers is not in the document]
Comment (MartinBasti) [reply atomic_reactor/source.py]: If autorebuild section in yaml is empty: 
```
---
autorebuild:
```
then the key `autorebuild` is defined with value `None` and `.get('autorebuild', {})` returns `None` instead `{}`. Functions later expect dict and they will fail with `None`.

```python
>>> print(dict(autorebuild=None).get('autorebuild', {}))
None
```

**Comment by lcarva** [comment]
Updated release notes draft:
> - empty sections allowed in `container.yaml`

**Pull request merged**